br/pkg/restore/data.go

```diff
@@ -140,138 +149,118 @@ func getStoreAddress(allStores []*metapb.Store, storeId uint64) string {
 }
 
 // read all region meta from tikvs
-func (recovery *Recovery) ReadRegionMeta(ctx context.Context, totalTiKVs int, allStores []*metapb.Store) error {
+func (recovery *Recovery) ReadRegionMeta(ctx context.Context) error {
 	eg, ectx := errgroup.WithContext(ctx)
-	workers := utils.NewWorkerPool(uint(totalTiKVs), "Collect Region Meta")
+	totalStores := len(recovery.allStores)
+	workers := utils.NewWorkerPool(uint(mathutil.Min(totalStores, max_store_concurency)), "Collect Region Meta") // TODO: int overflow?
 
 	// TODO: optimize the ErroGroup when TiKV is panic
-	metaChan := make(chan RecoveryMeta, 1024)
+	metaChan := make(chan StoreMeta, 1024)
 	defer close(metaChan)
 
-	for i := 0; i < totalTiKVs; i++ {
+	for i := 0; i < totalStores; i++ {
 		i := i
-		storeId := allStores[i].GetId()
-		tikvClient, err := recovery.newTiKVRecoveryClient(ectx, allStores[i].GetAddress())
+		storeId := recovery.allStores[i].GetId()
+		storeAddr := recovery.allStores[i].GetAddress()
+
+		if err := ectx.Err(); err != nil {
+			return errors.Trace(err)
+		}
+
+		tikvClient, conn, err := recovery.newTiKVRecoveryClient(ectx, storeAddr)
```

Comment:
if new client failed, previous running workers will be left running.

below there's 2 cases too